### ISSUE TITLE
Scene is not rendering

### ISSUE THREAD

**Issue by pooncs** [nerfstudio-project/nerfstudio#765]
Hi this is a really cool work.

I was testing this repo and was using the default scene in the quickstart page. Everything loaded correctly and I can see it training, however, when i open the viewer, I can only see the camera poses with the images. I am running this code on a AWS instance (ubuntu 20.04) terminal, with port 7007 opened. I can see all the stats updating, but the scene is not rendering. Thanks!
![Screen Shot 2022-10-12 at 5 07 58 PM](https://user-images.githubusercontent.com/22141463/195448807-36b1d08c-3615-4105-997c-0fb3dac12e61.png)

**Comment by eldercrow** [comment]
Same problem here; tested with blender dataset as well.

**Comment by ethanweber** [comment]
@eldercrow can you describe your machine setup as well? It’s something with webrtc not connecting.

**Comment by shaoguowen** [comment]
Same problem

**Comment by Blueslikehail** [comment]
Same problem for me. clean env but I am using cuda 11.7

**Comment by ethanweber** [comment]
Does refreshing after the images first show up, then waiting ~10 seconds, work for any of you? Curious if it’s due to the initialization sequence or because webrtc is failing with your machine setup.

**Comment by tancik** [comment]
Everyone who is running into this issue, can you report the following details (we are having trouble replicating this bug on the dev machines).

1. OS version
2. nerfstudio version
3. Browser type
4. Local or remote compute
5. What is your internet setup, (ie are you on a corp network, or personal plan)

Other tips:
+ Make sure to open link reported in browser, don't use an old one
+ Only have one viewer window open
+ Try refreshing

**Comment by tancik** [comment]
@Blueslikehail Can you try the latest version of the code to see if that fixes the issue. If it doesn't a few followup questions,

+ Are you are able to see the dataset images, but not the NeRF rendering?
+ Is this a new issue, or have previous versions of nerfstudio also not work?

I'm concerned that it may be caused by webrtc being blocked by some corporate network restrictions.

**Comment by Blueslikehail** [comment]
Hi Matthew,Just tried the last version and I have the same problem. Tomorrow I’ll try from within the corporate network to see if it works. Which previous version do you suggest trying should the problem persist?Thanks,WalterOn Oct 17, 2022, at 19:05, Matthew Tancik ***@***.***> wrote:﻿
@Blueslikehail Can you try the latest version of the code to see if that fixes the issue. If it doesn't a few followup questions,

Are you are able to see the dataset images, but not the NeRF rendering?
Is this a new issue, or have previous versions of nerfstudio also not work?

I'm concerned that it may be caused by webrtc being blocked by some corporate network restrictions.

**Comment by pooncs** [comment]
Ubuntu 20.04 (x86_64)
0.1.5.1
Edge
Remote
Remote on corporate site

I've tried the localhost too.
It gave me this address:
(https://viewer.nerf.studio/versions/22-10-13-0/?websocket_url=ws://localhost:7007)
which did not work too.

**Comment by tancik** [comment]
Just checking, everyone in this thread is able to see the dataset images, just not the render (same as in the image in the original post)?

**Comment by shaoguowen** [comment]
Ubuntu 18.04.4 LTS (GNU/Linux 5.4.0-126-generic)
0.1.5
Chrome
Remote
Remote on corporate site

Yes, I can see the  the dataset images, but no render scene.

**Comment by eldercrow** [comment]
Ubuntu 18.04.4 LTS 
Latest version (as of today)
NVIDIA V100
Edge
Remote
Remote on corporate site

Yes, I can see the the dataset images, but no render scene.

**Comment by Blueslikehail** [comment]
Hi All,I am running the same setup on corporate site and I can see the render now.WalterOn Oct 18, 2022, at 08:15, Hyunjoon Lee ***@***.***> wrote:﻿
Ubuntu 18.04.4 LTS
Latest version (as of today)
NVIDIA V100
Edge
Remote
Remote on corporate site
Yes, I can see the the dataset images, but no render scene.

**Comment by eldercrow** [comment]
@Blueslikehail Could you please let me know your setting? Has anything changed?

**Comment by Blueslikehail** [comment]
Hi,No changes, it seems I just have to be within the corporate network. WalterOn Oct 18, 2022, at 09:41, Hyunjoon Lee ***@***.***> wrote:﻿
@Blueslikehail Could you please let me know your setting? Has anything changed?

**Comment by tancik** [comment]
Sorry this is proving to be such an issue. It seems that webrtc initial connection is sometimes flaky. We we continue to try to debug this, unfortunately it isn't something we are experiencing on our machines.

(side note, we have been using chrome, it might be worth seeing if it works better on chrome)

**Comment by AhmetCanSolak** [comment]
I like to offer help, yet I am not able to reproduce it when I want it.

Btw, I updated my Firefox and not experiencing the issue anymore. In case it helps anyone out there. Also haven't experienced this issue on Chrome(windows 10) as well.

**Comment by libeanim** [comment]
Hi @tancik,
Thanks for your work!

Have the same issue when running `ns-train` on the docker image `nvidia/cuda:11.8.0-cudnn8-devel-ubuntu22.04` and ssh port forwarding.

![Screenshot from 2022-10-20 14-08-16](https://user-images.githubusercontent.com/1412303/196945055-36241798-d1d5-4881-a066-dab06f5f01ad.png)
Refreshing after a few training steps and then waiting does not help.
The RTC connection seems to be fine and Server + Render are shown as connected.
I've tested on Google Chrome and Firefox.

In the terminal everything seems to be normal and the `Vis Rays / Sec` column (which I assume is responsible for the viewer) also shows entries.
![Screenshot from 2022-10-20 14-07-34](https://user-images.githubusercontent.com/1412303/196945584-11e964fa-550b-4f4b-b2d2-7df43937857a.png)

**Comment by snarb** [comment]
I have the same issue when using port forwarding on the remote machine. When connecting locally, the scene is rendered correctly.

**Comment by arcticfly** [comment]
I had to reboot the remote Ubuntu machine I was working on.

OS version: OSX 12.5.1
nerfstudio version: 
Browser type: Chrome
Local or remote compute: remote
Internet setup: Personal plan

**Comment by peabody124** [comment]
Configurations that don't work for me:

OS version: OSX 12.6 (local OS, remote Ubuntu 20.04)
nerfstudio version: latest (as of today, also didn't work with 0.1.5)
Browser type: Chrome (and Firefox)
Local or remote compute: remote
Internet setup: Corporate internet

OS version: Ubuntu 20.04 (local OS, remote Ubuntu 20.04)
nerfstudio version: latest (as of today, also didn't work with 0.1.5)
Browser type: Firefox
Local or remote compute: remote
Internet setup: Corporate internet

Does work:
Remotely connected to the same machine with NoMachine using Firefox

**Comment by sharp-trickster** [comment]
Same issue here, but, weirdly, another dataset I trained loaded the view correcly.

Win 10
nerfstudio v0.1.8
Chrome 107.0.5304.88 (64 bits)
Run Local,
Viewing from localhost, so, no ISP issue.

**Comment by xiaohulihutu** [comment]
Same issue
I was working on Windows locally and it was fine.

But on the remote server and forward to windows, there is no render.

**Comment by kevinddchen** [comment]
I just got this issue again as well. 

Is it a cooncidence we have it at the same time? 👀

**Comment by naruya** [comment]
Same issue 😭
It used to work (few weeks? ago), but now it doesn't.

Win 11 + WSL (Ubuntu 20.04.5)
nerfstudio v0.1.8
Chrome, Firefox, Edge
Run Local,
Viewing from localhost

**Comment by xiaohulihutu** [comment]
> I just got this issue again as well.
> 
> Is it a cooncidence we have it at the same time? 👀

Now it makes sense. I checked locally still no render. I guess its something wrong with the render itself. Not our browser or setting issue. It used to be render connected, but now the render is disconnected all the time.

**Comment by tancik** [comment]
It's not a coincidence, mine is also down. It is because the TURN webRTC server we are using (https://www.metered.ca/tools/openrelay/) is currently down / overloaded. Hopefully it gets restored soon.

**Comment by sharp-trickster** [comment]
Wait: You need some kind of webservice even if you are running locally?
When you are viewing the localhost link and the connection is unstable, are you going through some remote server?!

**Comment by paulmelis** [comment]
> Sorry this is proving to be such an issue. It seems that webrtc initial connection is sometimes flaky. We we continue to try to debug this, unfortunately it isn't something we are experiencing on our machines.

Just ran into this as well, "Server connected" (seeing the image cameras in the viewer), but "Render Disconnected" (no rendered view). I have `ns-train` running on a remote GPU node and use an SSH tunnel to make remote port 7007 available locally, and 
then have the viewer connect to localhost:7007. Looking at some of the replies above it seems metered.ca is used to get around any firewall issues for the separate webRTC connection, in the browser console I see a `WebRTC: ICE failed`:

![image](https://user-images.githubusercontent.com/1408287/201545973-15434c8b-0818-4cea-bcbf-c737b588a13d.png)

Not sure what's going on there, but I suspect that one of the issues of why rendering doesn't work in my (and other) cases. Attaching logs ([aboutWebrtc.zip](https://github.com/nerfstudio-project/nerfstudio/files/9998293/aboutWebrtc.zip)) from Firefox's `about:webrtc`, maybe they provide some insight.

Client OS: Arch Linux, Firefox 106.0 (same issue in Chromium 106.0)
Server OS: Red Hat, nerfstudio 0.1.9
Connecting from home to remote HPC system through cable modem and SSHl tunnel

**Comment by tancik** [comment]
Thanks for the log! If anyone else has logs like this, please share. Also interested in similar logs in chrome.

**Comment by nlml** [comment]
I ran into this problem, but only if running nerfstudio inside a docker container. Works fine over ssh on a remote machine if not running in docker, but if I run in docker, I get the same problem as above.

**Comment by paulmelis** [comment]
One weird thing (might be a separate issue) is that even though I get correct rendering when running both training and web viewer on the same system, I still get a red `Render Disconnected` in the web UI.

**Comment by devernay** [comment]
I had the same issue when running the code on an EC2 instance: viewer either didn't show the nerf render, or showed it with a very low framerate.

**I got a 100% working solution** by using XRDP on the instance, launching a browser on the instance with the viewer URL, and connecting via RDP (using Microsoft Remote Desktop because I'm on Mac, but I guess there are linux solutions that work too) through a ssh tunnel.

Setup the ssh tunnel to the remote instance (`[::1]` worked for me, `localhost` didn't, because the XRDP server listened only on the IPv6 address, but `localhost` may work for you) 
```
ssh -N -L '12345:[::1]:3389' username@remote_host_name
```
then connect to localhost:12345 with MS Remote Desktop

**Comment by tancik** [comment]
Still seems to be an issue... :/

**Comment by olestole** [comment]
1. Ubuntu 20.04.5 LTS
2. Latest
3. Tried both Firefox and Chrome
4. Remote
5. University network

**I've tried all the tips to no avail.**

Nerfstudio has been working great for me until ~2 days ago. Now I only see the images as reported above, even if I'm running on a clean env and reset my local version to reflect origin/main. Any obvious reason why this suddenly started failing?

---

_When I try to only visualize a trained NeRF, I now get `STUN transaction timed out` after ~20 sec._

**Comment by penguinflys** [comment]
Seems I am not the only one having the issues😂, WebUI was okay previously, and somehow unable to connect to the render. 

BTW, thanks for the awesome codes and contributions on NeRFs 
![image](https://user-images.githubusercontent.com/29662741/208631985-c152e294-de5f-4d56-9e31-30389443c53a.png)

**Comment by tancik** [comment]
Try waiting a couple of minutes. Also try adding `--viewer.skip-openrelay` to the command.

**Comment by Miaosheng1** [comment]
I have the same problem 
![e55f76a990dff58fdea1b123fbb75ee](https://user-images.githubusercontent.com/111415805/209822349-b5b37abe-4881-4e5c-97c4-c3beacadb999.png)

**Comment by T-Photoworks** [comment]
I've been like this since yesterday. windows10,cuda11.6,python3.8
I've reinstalled about 20 times and it still doesn't work
![nerf](https://user-images.githubusercontent.com/25383398/210316941-e6cfaf71-974e-494f-bad3-382aec42effa.PNG)

**Comment by tancik** [comment]
The issues is networking related, not an installation issue.

Three potential workarounds:
+ Wait a minute or two, sometime connection is slow to initiate
+ Connect to a different network (or turn off VPN)
+ Try adding `--viewer.skip-openrelay` to command.

If anyone is experienced with webRTC and has an idea for a more permanent fix, please let us know.

**Comment by Avuerro** [comment]
I had the same problem where the renderer could not connect. After debugging it a bit I have found the following error 
```TypeError: initializer for ctype 'X509 *' must be a cdata pointer, not builtins.Certificate```.

This is triggered by the assertion on line 206 in rtcdtlstransport.py of the aiortc package. Fortunately, this issue has been reported on the aiortc repo, see https://github.com/aiortc/aiortc/issues/804. The proposed solution for now is to use cryptography package version <39. I'm using version 38 and the renderer currently works flawlessly for me. Hope this helps!

**Comment by T-Photoworks** [comment]
> 
Hi,
I have tried everything and the poster's scene does not appear. I gave it a few minutes of time. I don't use a VPN and I have disconnected to the internet and connected with my iPhone. 
Also, is this command "--viewer.skip-openrelay" to be added after "ns-train nerfacto --data data/nerfstudio/poster"? 
 I have tried but it gives me an error.
In the attached image, "Render Disconnected" is red, which is not normal?Does it matter?

**Comment by T-Photoworks** [comment]
> The issues is networking related, not an installation issue.
> 
> Three potential workarounds:
> 
> * Wait a minute or two, sometime connection is slow to initiate
> * Connect to a different network (or turn off VPN)
> * Try adding `--viewer.skip-openrelay` to command.
> 
> If anyone is experienced with webRTC and has an idea for a more permanent fix, please let us know.

Hi,
I have tried everything and the poster's scene does not appear. I gave it a few minutes of time. I don't use a VPN and I have disconnected to the internet and connected with my iPhone. 
Also, is this command "--viewer.skip-openrelay" to be added after "ns-train nerfacto --data data/nerfstudio/poster"? 
 I have tried but it gives me an error.
In the attached image, "Render Disconnected" is red, which is not normal?Does it matter?
![230104](https://user-images.githubusercontent.com/25383398/210512221-b657200d-15b8-42b6-b26b-b213e0407715.PNG)

**Comment by T-Photoworks** [comment]
> I had the same problem where the renderer could not connect. After debugging it a bit I have found the following error `TypeError: initializer for ctype 'X509 *' must be a cdata pointer, not builtins.Certificate`.
> 
> This is triggered by the assertion on line 206 in rtcdtlstransport.py of the aiortc package. Fortunately, this issue has been reported on the aiortc repo, see [aiortc/aiortc#804](https://github.com/aiortc/aiortc/issues/804). The proposed solution for now is to use cryptography package version <39. I'm using version 38 and the renderer currently works flawlessly for me. Hope this helps!

sorry. My knowledge does not allow me to understand your explanation.

**Comment by T-Photoworks** [comment]
> The issues is networking related, not an installation issue.
> 
> Three potential workarounds:
> 
> * Wait a minute or two, sometime connection is slow to initiate
> * Connect to a different network (or turn off VPN)
> * Try adding `--viewer.skip-openrelay` to command.
> 
> If anyone is experienced with webRTC and has an idea for a more permanent fix, please let us know.

I tried outputting the camera path without appear the scene.
However
”The viewer bridge server subprocess failed.”
'Try modifying --viewer.websocket-port 7007'
An error will occur with a message like this.
![230104C](https://user-images.githubusercontent.com/25383398/210515995-cd085bf4-0898-4dae-a357-6d0abad3c583.PNG)

**Comment by EvgenyKashin** [comment]
1. Ubuntu 18.04.6 LTS
2. v0.1.14
3. Chrome
4. Remote
5. University network

**Downgrading** **cryptography** package solved the issue for me:
```
pip uninstall -y cryptography
pip install cryptography==38
```

My error before was as follows:
```
_openssl_assert(lib.SSL_CTX_use_certificate(ctx, self._cert._x509) == 1)  # type: ignore AttributeError: 'builtins.Certificate' object has no attribute '_x509'
```
Found solution here: 
https://github.com/aiortc/aiortc/issues/804#issuecomment-1369182326

**Comment by T-Photoworks** [comment]
In my case, It fixed it by installing nerfstudio several times. I wish I could tell you why, but I don't know.thank you all.

**Comment by MikePelton** [comment]
Can confirm @EvgenyKashin's solution (downgrading the "cryptography" package to version 38) fixed this for me on Ubuntu 22.04 running locally - thanks Evgenii!

**Comment by qhdqhd** [comment]
> 1. Ubuntu 18.04.6 长期支持版
> 2. v0.1.14
> 3. 铬合金
> 4. 偏僻的
> 5. 大学网
> 
> **降级** **加密**包为我解决了这个问题：
> 
> ```
> pip uninstall -y cryptography
> pip install cryptography==38
> ```
> 
> 我之前的错误如下：
> 
> ```
> _openssl_assert(lib.SSL_CTX_use_certificate(ctx, self._cert._x509) == 1)  # type: ignore AttributeError: 'builtins.Certificate' object has no attribute '_x509'
> ```
> 
> 在这里找到解决方案： [aiortc/aiortc#804（评论）](https://github.com/aiortc/aiortc/issues/804#issuecomment-1369182326)

Thanks!
It solves my issue!

**Comment by SamedAli** [comment]
> 1. Ubuntu 18.04.6 LTS
> 
>     2. v0.1.14
> 
>     3. Chrome
> 
>     4. Remote
> 
>     5. University network
> 
> 
> **Downgrading** **cryptography** package solved the issue for me:
> 
> ```
> pip uninstall -y cryptography
> pip install cryptography==38
> ```
> 
> My error before was as follows:
> 
> ```
> _openssl_assert(lib.SSL_CTX_use_certificate(ctx, self._cert._x509) == 1)  # type: ignore AttributeError: 'builtins.Certificate' object has no attribute '_x509'
> ```
> 
> Found solution here: [aiortc/aiortc#804 (comment)](https://github.com/aiortc/aiortc/issues/804#issuecomment-1369182326)

FYI: This also worked for me.

**Comment by arielbenitah** [comment]
> 1. Ubuntu 18.04.6 LTS
> 
>     2. v0.1.14
> 
>     3. Chrome
> 
>     4. Remote
> 
>     5. University network
> 
> 
> **Downgrading** **cryptography** package solved the issue for me:
> 
> ```
> pip uninstall -y cryptography
> pip install cryptography==38
> ```
> 
> My error before was as follows:
> 
> ```
> _openssl_assert(lib.SSL_CTX_use_certificate(ctx, self._cert._x509) == 1)  # type: ignore AttributeError: 'builtins.Certificate' object has no attribute '_x509'
> ```
> 
> Found solution here: [aiortc/aiortc#804 (comment)](https://github.com/aiortc/aiortc/issues/804#issuecomment-1369182326)

Worked for me. Thanks a lot!

**Comment by tancik** [comment]
If upgrading cryptography doesn't work, try the following:

> The issues is networking related, not an installation issue.
> 
> Three potential workarounds:
> 
> * Wait a minute or two, sometime connection is slow to initiate
> * Connect to a different network (or turn off VPN)
> * Try adding `--viewer.skip-openrelay` to command.
> 
> If anyone is experienced with webRTC and has an idea for a more permanent fix, please let us know.

**Comment by DanWuh** [comment]
Thanks for all the assistance folks, confirming that ;

```
pip uninstall -y cryptography
pip install cryptography==38
```

along with changing the websocket to 7007 per the terminal recommendation `( --viewer.websocket-port 7007)` worked a treat for me.  Cheers!

**Comment by KamiCalcium** [comment]
> I ran into this problem, but only if running nerfstudio inside a docker container. Works fine over ssh on a remote machine if not running in docker, but if I run in docker, I get the same problem as above.
> 
> The command I am running is `ns-train nerfacto --vis viewer --viewer.websocket-port 8844`
> 
> And yes I am forwarding the port out of Docker ( `docker run --gpus all --ipc host -it --rm -v /home/me:/home/me -p 8844:8844 nerfstudio`)
> 
> Here is what it looks like when I run outside of Docker: ![good](https://user-images.githubusercontent.com/12813713/202664804-2221ad00-1ad4-424c-87f7-53b412462959.png)
> 
> And here is what I get if I run within Docker (same command) ![bad](https://user-images.githubusercontent.com/12813713/202665110-d5a5b6c5-d797-4f1f-8735-87ee40291399.png)
> 
> Here is the Dockerfile I used to construct the nerfstudio docker container - perhaps the error can be reproduced by others + fixed? Note the build command I used at the end of the Dockerfile.
> 
> ```
> FROM nvidia/cuda:11.6.0-devel-ubuntu20.04
> 
> # Prevent stop building ubuntu at time zone selection.  
> ENV DEBIAN_FRONTEND=noninteractive
> 
> # Prepare and empty machine for building
> RUN apt-get update && apt-get install -y \
>     git \
>     cmake \
>     build-essential \
>     libboost-program-options-dev \
>     libboost-filesystem-dev \
>     libboost-graph-dev \
>     libboost-system-dev \
>     libboost-test-dev \
>     libeigen3-dev \
>     libsuitesparse-dev \
>     libfreeimage-dev \
>     libmetis-dev \
>     libgoogle-glog-dev \
>     libgflags-dev \
>     libglew-dev \
>     qtbase5-dev \
>     libqt5opengl5-dev \
>     libcgal-dev \
>     libmetis-dev
> 
> # Build and install ceres solver
> RUN apt-get -y install \
>     libatlas-base-dev \
>     libsuitesparse-dev
> ARG CERES_SOLVER_VERSION=2.1.0
> RUN git clone https://github.com/ceres-solver/ceres-solver.git --branch ${CERES_SOLVER_VERSION}
> RUN cd ceres-solver && \
> 	mkdir build && \
> 	cd build && \
> 	cmake .. -DBUILD_TESTING=OFF -DBUILD_EXAMPLES=OFF && \
> 	make -j4 && \
> 	make install
> 
> # Build and install COLMAP
> 
> # Note: This Dockerfile has been tested using COLMAP pre-release 3.7.
> # Later versions of COLMAP (which will be automatically cloned as default) may
> # have problems using the environment described thus far. If you encounter
> # problems and want to install the tested release, then uncomment the branch
> # specification in the line below
> RUN git clone https://github.com/colmap/colmap.git #--branch 3.7
> 
> RUN cd colmap && \
> 	git checkout dev && \
> 	mkdir build && \
> 	cd build && \
> 	cmake .. && \
> 	make -j4 && \
> 	make install
> 
> RUN apt-get install -y wget
> 
> RUN INSTALL_PATH=~/anaconda \
>     && wget http://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh \
>     && bash Miniconda3-latest* -fbp $INSTALL_PATH
> 
> ENV PATH=/root/anaconda/bin:$PATH
> 
> RUN conda create --name nerfstudio -y python=3.8
> 
> SHELL ["conda", "run", "-n", "nerfstudio", "/bin/bash", "-c"]
> 
> RUN pip install --upgrade pip
> RUN pip install --no-cache torch==1.12.1+cu116 torchvision==0.13.1+cu116 -f https://download.pytorch.org/whl/torch_stable.html
> 
> ENV PATH=/usr/local/cuda/bin:$PATH
> ENV CUDA_HOME="/usr/local/cuda-11.6/"
> ARG TCNN_CUDA_ARCHITECTURES="86"
> 
> RUN pip install git+https://github.com/NVlabs/tiny-cuda-nn/#subdirectory=bindings/torch
> RUN pip install nerfstudio
> 
> RUN . /root/.bashrc && conda init bash
> 
> RUN echo "conda activate nerfstudio" >> ~/.bashrc
> 
> RUN apt-get install -y ffmpeg
> 
> # docker build -t nerfstudio --build-arg TCNN_CUDA_ARCHITECTURES=$(python -c "import torch; ma, m = torch.cuda.get_device_capability(); print(ma * 10 + m, end='')") .
> ```

Similar here. Local works but docker does not work. How did you solve it? My cryptography is 38.0.4 already.

**Comment by Holmes-Alan** [comment]
> Same issue 😭 It used to work (few weeks? ago), but now it doesn't.
> 
> Win 11 + WSL (Ubuntu 20.04.5) nerfstudio v0.1.8 Chrome, Firefox, Edge Run Local, Viewing from localhost

Same for me. It worked before and now it does not work, do you have solutions?

**Comment by KamiCalcium** [comment]
> > Same issue 😭 It used to work (few weeks? ago), but now it doesn't.
> > Win 11 + WSL (Ubuntu 20.04.5) nerfstudio v0.1.8 Chrome, Firefox, Edge Run Local, Viewing from localhost
> 
> Same for me. It worked before and now it does not work, do you have solutions?

Local or Docker? Local works for me. Try update to nerfstudio 0.1.15. Sometimes you need to wait and refresh. Docker still does not work for me but it might be my own network issue.

**Comment by Holmes-Alan** [comment]
> > > Same issue 😭 It used to work (few weeks? ago), but now it doesn't.
> > > Win 11 + WSL (Ubuntu 20.04.5) nerfstudio v0.1.8 Chrome, Firefox, Edge Run Local, Viewing from localhost
> > 
> > 
> > Same for me. It worked before and now it does not work, do you have solutions?
> 
> Local or Docker? Local works for me. Try update to nerfstudio 0.1.15. Sometimes you need to wait and refresh. Docker still does not work for me but it might be my own network issue.

Thanks, I resolved it by waiting for a few seconds. I can get it work in docker by assigning a unique port number, like --viewer.websocket-port 80

**Comment by chris838** [comment]
Just to add my 2c - I had nothing rendered and the red "Render Disconnected". I think clicking "Hide Images" seemed to help. Also pausing training, refreshing the Chrome browser page, then resuming. And not opening up multiple viewer instances.

**Comment by jdiazram** [comment]
I was the same problem, but I changed the connection to the internet with my cell phone connection, and it works!

**Comment by LemonZhong** [comment]
yeah, it really works with the command
pip uninstall -y cryptography
pip install cryptography==38

**Comment by machenmusik** [comment]
There was a PR to use cryptography 38 (until that gets fixed), should be in latest unless undone

**Comment by Sandros94** [comment]
> I can get it work in docker by assigning a unique port number, like --viewer.websocket-port 80

Same here, I was able to connect to the render, via docker, only if I specified the --viewer.websocket-port 7007 (idk what is the default port, but surely not 7007).

Windows 11
Docker Desktop
WSL2
0.1.16

**Comment by dooonnis** [comment]
Same here as well, my viewer also only shows "Render disconnected".
I am connected to the server with the viewer via port 7007 and I'm actually able create videos of a trained model using blindly created camera paths. But I'm just not getting any live feedback of the NeRF model from inside the viewer.

None of the above solutions have worked for me so far, although I believe there are multiple problems adressed in this thread.

Neither using cryptography 38 nor _-viewer.websocket-port 7007_ and -_viewer.skip-openrelay_ helped.

Windows 10
Docker Desktop
WSL2
0.1.16

**Comment by dbernatt** [comment]
Same here. I was unable to render anything in the viewer. I tried to fix it with several suggestions mentioned in this thread, but with no luck at all. I'm running the colab version.

Chrome console output: 
![image](https://user-images.githubusercontent.com/72317320/216523796-cf7a0bdd-dad0-45eb-9486-1caa2f41d648.png)

**Comment by tancik** [comment]
@dbernatt I think your error is different, it looks like the socket isn't connecting (not a webrtc error). We use localtunnel to create a port, but for some reason that seems to have failed for you.

**Comment by dooonnis** [comment]
I also ran into this issue. However, I found a few things out. 

Using conda:
The NeRF never showed up and the viewer always showed "Render disconnected". I then used a different network (I used a mobile hotspot and reloaded the page) and then the NeRF model instantly showed up. Since it's repeatable, it's definitely something network in this case.

Comparing the browser console logs, it seems to be something WebRTC related, because when connected to the "faulty" network, WebRTC stops after the point "sending offer", while in the other network, WebRTC receives the needed answer and then renders the NeRF, while it still shows "Render disconnected".

Using Docker with WSL 2:
Unfortunately, with Docker this is not solved so easily. With the "faulty" network there is still no answer being received, while in the other network there is, however there is still nothing being rendered. There I am receiving the following error messaged in the terminal:

Task exception was never retrieved
future: <Task finished name='Task-42' coro=<TurnClientMixin.send_data() done, defined at /home/user/.local/lib/python3.10/site-packages/aioice/turn.py:244> exception=TransactionFailed(Message(message_method=Method.CHANNEL_BIND, message_class=Class.ERROR, transaction_id=b'\xa7=\xf8\xd4\xdb?/\xaePz\xef\x91'))>
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 263, in send_data
    await self.channel_bind(channel, addr)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 88, in channel_bind
    await self.request_with_retry(request)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 216, in request_with_retry
    response, addr = await self.request(request)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 203, in request
    return await transaction.run()
  File "/home/user/.local/lib/python3.10/site-packages/aioice/stun.py", line 299, in run
    return await self.__future
aioice.stun.TransactionFailed: STUN transaction failed (403 - Forbidden IP)
Task exception was never retrieved
future: <Task finished name='Task-44' coro=<TurnClientMixin.send_data() done, defined at /home/user/.local/lib/python3.10/site-packages/aioice/turn.py:244> exception=TransactionFailed(Message(message_method=Method.CHANNEL_BIND, message_class=Class.ERROR, transaction_id=b':\xc8\x9a\xa8\x08VF\x17\x05\xffh;'))>
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 263, in send_data
    await self.channel_bind(channel, addr)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 88, in channel_bind
    await self.request_with_retry(request)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 216, in request_with_retry
    response, addr = await self.request(request)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 203, in request
    return await transaction.run()
  File "/home/user/.local/lib/python3.10/site-packages/aioice/stun.py", line 299, in run
    return await self.__future
aioice.stun.TransactionFailed: STUN transaction failed (403 - Forbidden IP)
Task exception was never retrieved
future: <Task finished name='Task-47' coro=<TurnClientMixin.send_data() done, defined at /home/user/.local/lib/python3.10/site-packages/aioice/turn.py:244> exception=TransactionFailed(Message(message_method=Method.CHANNEL_BIND, message_class=Class.ERROR, transaction_id=b'\x90v&\xab\x12\x8d\xde\xf3\x83\xac%\x04'))>
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 263, in send_data
    await self.channel_bind(channel, addr)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 88, in channel_bind
    await self.request_with_retry(request)
  File "/home/user/.local/lib/python3.10/site-packages/aioice/turn.py", line 216, in request_with_retry
    response, addr = await self.request(request)

I'm using Windows 10.

**Comment by dooonnis** [comment]
I just found the solution in this Issue #1042.

Adding `--viewer.skip-overlay True` to the command line solves the issue of the NeRF not showing up both when using Conda and Docker.

The command would then look like [this](https://github.com/nerfstudio-project/nerfstudio/pull/1042#issuecomment-1411554122):
`ns-train nerfacto --data DATA --viewer.skip-openrelay True`

**Comment by braincharter** [comment]
> I just found the solution in this Issue #1042.
> 
> Adding `--viewer.skip-overlay True` to the command line solves the issue of the NeRF not showing up both when using Conda and Docker.
> 
> The command would then look like [this](https://github.com/nerfstudio-project/nerfstudio/pull/1042#issuecomment-1411554122): `ns-train nerfacto --data DATA --viewer.skip-openrelay True`

It did not work for me, I still have "Render Disconnected"
(I am on WSL2)
I tried reinstalling cryptography 38 as mentionned earlier, but I do have this error that pops up in the terminal:

`key=crypto.PKey.from_cryptography_key(key),
  File "/home/.../miniconda3/envs/py39-wsl/lib/python3.9/site-packages/OpenSSL/crypto.py", line 274, in from_cryptography_key
    raise TypeError("Unsupported key type")`

**Comment by tancik** [comment]
Updated workarounds

+ Wait a minute or two, sometime connection is slow to initiate
+ Connect to a different network (or turn off VPN)
+ Try adding `--viewer.skip-openrelay True` to command.
+ Try adding `--viewer.local True --viewer.skip-openrelay False` to command.

If anyone is experienced with webRTC and has an idea for a more permanent fix, please let us know.

**Comment by pacifinapacific** [comment]
Has anyone had success with port fording to a remote machine in docker?

**Comment by XinyangHan** [comment]
I am using a remote server connected using ssh and vscode as editor. I successfully go through other operations in the quick start, and I can observe the port forward being automatically launched in the vscode. But when I open the browser with the url locally, there is nothing being rendered. (Not training images or the rendering scene)

![1676780442675_720](https://user-images.githubusercontent.com/86959032/219922098-327b948b-7c30-4ab6-980f-7c4858c08264.png)
![1676780496536_720](https://user-images.githubusercontent.com/86959032/219922100-943a7404-d312-4122-bd91-68c1730b23ab.png)

**Comment by apollo600** [comment]
I ran into the same problem too, and I solved the problem by install nerfstudio through cloning the repository `https://github.com/nerfstudio-project/nerfstudio.git`, so that the package `nerfstudio` will be the latest and the greatest

```bash
git clone https://github.com/nerfstudio-project/nerfstudio.git
cd nerfstudio
pip install --upgrade pip setuptools
pip install -e .
```

My Environment is:

- Windows 11
- CUDA 11.7
- functorch 1.13.1
- tinycudann 1.7

The final viewer looks like below, it worked!😃 
![1~G1GM(N4%%N5WQ2{{LVY_3](https://user-images.githubusercontent.com/68414883/219963505-7321e906-afa5-4ddb-b362-5ec77f870b9b.jpg)

**Comment by kenchen3000** [comment]
Hi I have same issue with latest code.

OS version: Fedora
nerfstudio version: 1.18
Browser type: Chrome
Local 
Home internet

**Comment by yanggg1997** [comment]
I also encountered this problem, and have found the method to skip it for my condition:
**The cause of my problem is that I accessed the wrong address!**
My Edge&Chrome browsers remember the history url, so when I type "viewer.nerf..."，the browser will autocomplete the full url address I accessed last time. However, each time the correct url changes. What I need to do is copy and access the new url address printed in the terminal when I launch a ns-train task, or just access `https://viewer.nerf.studio/` also works.

**Comment by Mirrracle** [comment]
> 



> Updated workarounds
> 
> * Wait a minute or two, sometime connection is slow to initiate
> * Connect to a different network (or turn off VPN)
> * Try adding `--viewer.skip-openrelay True` to command.
> * Try adding `--viewer.local True --viewer.skip-openrelay False` to command.
> 
> If anyone is experienced with webRTC and has an idea for a more permanent fix, please let us know.

Hi! I can successfully train the model and render it export the video using ns-render, but just cannot see the render result from the viewer. I tried the two ways you mentioned but I don't know why it says unrecognized arguments. 
![image](https://user-images.githubusercontent.com/75305926/226168621-ed216d3a-64df-4d75-8f41-24f6d0a77ab2.png)

**Comment by tancik** [comment]
@YueYin27 We moved away from webrtc so those arguments are no longer applicable. Check that your port your are using is free.

**Comment by tancik** [comment]
@sandy-ssdut Update to the latest version of nerfstudio to solve this issue. We were previously relying on a free service for making the webrtc link. They have since changed their policies which is why it doesn't connect.

**Comment by Hiroaki-K4** [comment]
Hi. Thanks for the coolest job you guys do.
I have one question. When I am using docker in a windows 11 WSL2 environment and doing the model training, I get an error like the attached image in the viewer and cannot display it properly. Is there any solution to this problem?

![nerf](https://user-images.githubusercontent.com/51109408/234171791-d49a5f44-db9b-4c84-9b27-10a6a234ecff.png)

**Comment by orrblue** [comment]
Hi, I'm experiencing a similar issue where the scene is not rendering. The training images show successfully, but no scene is visible, including after adding a camera and waiting. No errors in console. Furthermore, I am seeing "iteration: 0" and "Resolution: ?" like in the screenshot above and some previous screenshots. Thank you.

My system:
* Chrome and Firefox
* Ubuntu 20.04
* Followed install instructions for Torch 2.0.1 with CUDA 11.8
* * But instead of `pip install nerfstudio` I did `pip install git+https://github.com/nerfstudio-project/nerfstudio.git` as I was advised by [this suggestion to fix a training issue](https://github.com/nerfstudio-project/nerfstudio/issues/724#issuecomment-1553781828) I had.

**Comment by orrblue** [comment]
Actually, it seems fixed for me now -- scene renders while training. I re-ran training and then I opened opened the viewer link only in one tab, and only in google chrome (as advised a while up this thread).

**Comment by flow-specter** [comment]
> 

How can we check if the render is disconneted?

**Comment by raja-kumar** [comment]
I have tried almost everything, but nothing seems to work. I installed it from the source. I have forwarded the traffic from the remote still it doesn't render the scene. I can see the camera positions and images. Is anyone still facing this issue, and have any idea how to resolve this?

**Comment by tancik** [comment]
@raja-kumar what version are you running? Can you provide a screenshot. The code for this was completely rewritten to avoid this issue so it's odd that you are still running into it.

**Comment by raja-kumar** [comment]
I have installed it from the source to use the latest dev version. (as instructed [here](https://docs.nerf.studio/en/latest/quickstart/installation.html#installing-nerfstudio)). See the screenshot below.
<img width="1427" alt="Screenshot 2023-06-22 at 5 44 57 PM" src="https://github.com/nerfstudio-project/nerfstudio/assets/26834297/aee5cb60-32ea-41d2-bb44-94be50e98607">

**Comment by tancik** [comment]
Does the training start in the terminal?

**Comment by raja-kumar** [comment]
> Does the training start in the terminal?

Yes it does. I even tried viewer after model is trained. It still doesn't show the scene

**Comment by tancik** [comment]
What browser are you using? Perhaps an extension is blocking it?

**Comment by raja-kumar** [comment]
> What browser are you using? Perhaps an extension is blocking it?

I am using Google Chrome as suggested in this thread. I tried in incognito mode still doesn't work. So probably not an issue because of the extension.

**Comment by tancik** [comment]
Can you open up the developer tools console in chrome and see if it has any errors. Also can you post the URL that you are navigating to.

**Comment by raja-kumar** [comment]
> Can you open up the developer tools console in chrome and see if it has any errors. Also can you post the URL that you are navigating to.

I do see an error in the console. 

Uncaught DOMException: Failed to execute 'send' on 'WebSocket': Still in CONNECTING state.
    at a (https://viewer.nerf.studio/versions/23-05-15-1/static/js/main.1402ae1c.js:2:1144453)
    at N (https://viewer.nerf.studio/versions/23-05-15-1/static/js/main.1402ae1c.js:2:1598884)
Viser connected! ws://localhost:7007 ViserWebSocket.tsx:359

here is the URL

https://viewer.nerf.studio/versions/23-05-15-1/?websocket_url=ws://localhost:7007

**Comment by tancik** [comment]
Im at a bit of a loss, I have no idea why this would be happening. One last suggestion is to re-run `pip install -e .`, perhaps your version of viser if out of date.

**Comment by raja-kumar** [comment]
> Im at a bit of a loss, I have no idea why this would be happening. One last suggestion is to re-run `pip install -e .`, perhaps your version of viser if out of date.

I re-run `pip install -e .` but still it doesn't work. anyways, Thank you for your prompt response. 

Are there some scripts I can use to render from different viewpoints using the trained model (rather than the viewer)? I tried looking up the documentation but couldn't find it.

**Comment by tancik** [comment]
You can run `--vis tensorboard` or `--vis wandb` to see eval images.
One last thing you could try, launch your own local viewer. You can find instructions [here](https://docs.nerf.studio/en/latest/developer_guides/viewer/local_viewer.html)

**Comment by raja-kumar** [comment]
@tancik I think I found the issue. The remote server I am using is on the university network, and the port forwarding is probably getting blocked due to the university's network security/ firewall. I tried it on another server of an outside network, and it's working perfectly fine now. Thank you for your help! Appreciate it.

**Comment by Eatonasher** [comment]
Mine won't connect to the renderer either, and I'm running on google colab. What was weird, was after 10 min of not refreshing, a few pictures popped up. Then, after I hit refresh, they disappeared and the "disconnected renderer" message resumed.

Curious eh? Not sure what the issue might be.

**Comment by guswhd** [comment]
Hi. thanks for your API. but I'm encounting some problem like other people struggled problem.
I install using docker(version: 0.1.18 with cuda 11.8)
it working fine at learning but not rendering at web page, I even made sure this is rendered as a video file.
please help me!!
![ns render disconnected](https://github.com/nerfstudio-project/nerfstudio/assets/114475395/d8c835e1-22f4-404c-bd6d-f44eb456da0b)

**Comment by kerrj** [comment]
@guswhd what is your training setup? make sure the network connection/port  isn't blocked somehow via firewall etc. Since you're in docker you might need to do something special to forward the port from the container? not sure, haven't used docker that much personally.

**Comment by guswhd** [comment]
@kerrj  I'm still an undergraduate research student, so I don't know much :(
Are you saying that network settings should be done within windows? I've tried disabling the Windows Firewall altogether, but the same problem persists.

**Comment by guswhd** [comment]
@kerrj Thank U for your answer and my problem resolved to update version to 0.3.0, rendering environment(?) exactly.

**Comment by danhussey** [comment]
In case it might help anyone, I was getting 'renderer disconnected' when accessing https://viewer.nerf.studio/versions/23-05-15-1/?websocket_url=ws://localhost:7007 on Safari, while running ns-train on an Azure VM. I had set up inbound networking rules to allow forwarding of port 7007, and confirmed the port was being forwarded with the -v flag while ssh'ing from my MacBook. 

Long story short: it worked when accessing it using Chrome, but not safari.

**Comment by nekoshadow1** [comment]
For me. I have tried all methods above, but nothing worked. Then I switched to a new machine, and it worked perfectly.
I'm still not sure what went wrong with the first machine. The only difference between them is the version of OS (Ubuntu 18.04 vs 22.04)

**Comment by yt-koike** [comment]
In my case, Firefox didn't work and Chrome did.
